notebooks/cloud-functions-template/notebook.ipynb

```diff
@@ -0,0 +1,305 @@
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "id": "e0f88a6f-2658-40b4-9356-935a09f5053e",
+      "metadata": {},
+      "source": [
+        "<div class=\"alert alert-block alert-warning\">\n",
+        "    <b class=\"fa fa-solid fa-exclamation-circle\"></b>\n",
+        "    <div>\n",
+        "        <p><b>Note</b></p>\n",
+        "        <p>This notebook can be run on a Free Starter Workspace. To create a Free Starter Workspace navigate to <tt>Start</tt> using the left nav. You can also use your existing Standard or Premium workspace with this Notebook.</p>\n",
+        "    </div>\n",
+        "</div>"
+      ]
+    },
+    {
+      "attachments": {},
+      "cell_type": "markdown",
+      "id": "a5564913-7ff8-41bf-b64b-b67971c63fae",
+      "metadata": {},
+      "source": [
+        "This Jupyter notebook will help you build your first Cloud Function, showcasing how to leverage the ultra-fast queries of SingleStore to build a responsive API server using FastAPI"
+      ]
+    },
+    {
+      "attachments": {},
+      "cell_type": "markdown",
+      "id": "1e394195-29b4-403c-9abf-5d7731349eb6",
+      "metadata": {},
+      "source": [
+        "## Create some simple tables\n",
+        "\n",
+        "This setup establishes a basic relational structure to store some items information."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 1,
+      "id": "a17bdd3a-16b3-4e19-8a56-6566a169eccb",
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "%%sql\n",
+        "DROP TABLE IF EXISTS items;\n",
+        "\n",
+        "CREATE TABLE IF NOT EXISTS\n",
+        "items (\n",
+        "    id INT PRIMARY KEY,\n",
+        "    name VARCHAR(255),\n",
+        "    price FLOAT\n",
+        ");"
+      ]
+    },
+    {
+      "attachments": {},
+      "cell_type": "markdown",
+      "id": "af6e2618-de97-4397-b0d2-23e4a4df1d83",
+      "metadata": {},
+      "source": [
+        "## Create a Connection Pool\n",
+        "\n",
+        "To run multiple simultaneous queries, we use sqlalchemy to create a pool of sql connections to the workspace you have selected. We also define a method to execute queries and transactions using a connection from this pool."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 2,
+      "id": "f485e71b-2b05-4696-b22a-cf046fd83090",
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "from sqlalchemy import create_engine, text\n",
+        "import requests\n",
+        "\n",
+        "ca_cert_url = \"https://portal.singlestore.com/static/ca/singlestore_bundle.pem\"\n",
+        "ca_cert_path = \"/tmp/singlestore_bundle.pem\"\n",
+        "\n",
+        "response = requests.get(ca_cert_url)\n",
+        "with open(ca_cert_path, \"wb\") as f:\n",
+        "    f.write(response.content)\n",
+        "\n",
+        "sql_connection_string = connection_url.replace(\"singlestoredb\", \"mysql+pymysql\")\n",
```

Comment:
Why are you pulling out the singlestoredb driver and using pymysql?

Reply:
This is a demo for dashboards and cloud functions.
%%sql doesnt work in a multi threaded manner, which is why we need to use the sqlalchemy engine with a connection pool for now.

Reply:
I still don't understand. Why doesn't SinlgeStoreDB work? PyMySQL has the exact same thread issues that the SingleStoreDB client does.